### PR TITLE
kv/bulk: fix initial splits

### PR DESCRIPTION
**Pull request by dt**
Previously initial splits were missing at the first key due to how the loop
initialized. This meant that a processor Y that was ingesting data in the 
span `[k, p)` would write keys `k`, `j`, `l`, etc but not split until, say key `m`. 
When another processor X was ingesting into the prior span `[a, k)`, it regularly
splits and scatters the "empty" RHS of its span before it fills some prefix of it and
then splits and scatters the remaining span again. However the fact that processor 
Y did not split at `k` meant the "empty" range containing the empty suffix of `[a, k)` 
actually was not an empty range and instead had `k`, `j`, etc in it, which then had 
to be actually moved every time X scattered that range. 

Now the first key is used as a split key to avoid this. This requires using a
different predicate key than usual; typically we check if where we are
splitting at is still in the same range as the prior split to detect if
another node has already split the span, but we have no prior split on
the first one. Instead, we use the next split for the first one's predicate
as this can also serve to show that the span is still 'wide' enough to split.

Release note: none.

Release justification: low-risk fix of new functionality.

### CODE REVIEW

**Comment by dt** [comment]
Only last commit is new, rest is #77399.

**Comment by dt** [comment]
@aayushshah15 @adityamaru Note that I don't think this fully explains the observed regression, since we first observed it before I "fixed" the initial-splits conditional; prior to https://github.com/cockroachdb/cockroach/pull/77352, I believe we never called createInitialSplits, so the sstBatcher's own split "if this is the first flush, add a split" logic should have been making this first split.

**Review by nvb — COMMENTED** [review]
Heh, this is exactly the kind of thing @aayushshah15 and I were speculating about when discussing these non-empty splits. Nice find. Have you run any tests to demonstrate the impact of this change?

> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77403#-:-Mx_4pGfA1rLQ0LfGxly:bd3rf88)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @adityamaru, and @dt)

___
*[-- commits, line 10 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77403#-Mx_4PeC1N3aPQOsSYJu):*
> Instead, we use the next split for the first one's predicate as this can also serve to show that the span is still 'wide' enough to split.

If the span is no longer wide enough to split, won't this lead to non-empty scatters? Did you consider not using a predicate key at all in this case?
___
*[pkg/kv/bulk/buffering_adder.go, line 370 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77403#-Mx_3jch7QbG7ZVhSEcM:-Mx_3jch7QbG7ZVhSEcN:b-ad8baa) ([raw file](https://github.com/cockroachdb/cockroach/blob/036c0a47dd123cf629a99ec3983373f318f15cc1/pkg/kv/bulk/buffering_adder.go#L370)):*
> ```Go
> 		}
> 		predicateAt := splitAt - width
> 		if predicateAt < 0 {
> ```

Could you add a comment explaining this case?
___
*[pkg/kv/bulk/buffering_adder.go, line 371 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77403#-Mx_3sMxBNdBsf1z0kzv:-Mx_3sMxBNdBsf1z0kzw:b-iybdp5) ([raw file](https://github.com/cockroachdb/cockroach/blob/036c0a47dd123cf629a99ec3983373f318f15cc1/pkg/kv/bulk/buffering_adder.go#L371)):*
> ```Go
> 		predicateAt := splitAt - width
> 		if predicateAt < 0 {
> 			if next := splitAt + width; next < len(b.curBuf.entries) {
> ```

nit: this would be easier to understand if this was written as:
```go
next := splitAt + width
if next > len(b.curBuf.entries) - 1 {
    next = len(b.curBuf.entries) - 1
}
```

**Review by nvb — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77403)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @adityamaru, and @dt)

___
*[pkg/kv/bulk/buffering_adder.go, line 380 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77403#-Mx_56v25HjSbTD_Klkv:-Mx_56v25HjSbTD_Klkw:bwi2771) ([raw file](https://github.com/cockroachdb/cockroach/blob/036c0a47dd123cf629a99ec3983373f318f15cc1/pkg/kv/bulk/buffering_adder.go#L380)):*
> ```Go
> 		predicateKey := b.curBuf.Key(predicateAt)
> 		log.VEventf(ctx, 1, "pre-splitting span %d of %d at %s", i, b.initialSplits, splitKey)
> 		resp, err := b.sink.db.SplitAndScatter(ctx, splitKey, hour, predicateKey)
> ```

Do you think we should be using a longer expiration for the first split in each partition?

**Review by dt — COMMENTED** [review]
> when discussing these non-empty splits

Note that as mentioned above though that prior to #77352 this was accidentally dead code, so this doesn't explain badness observed e.g. a week ago.

> Have you run any tests to demonstrate the impact of this change?

I've done some tests and wrote up some in an internal slack thread here: https://cockroachlabs.slack.com/archives/C2C5FKPPB/p1646519993744969?thread_ts=1646263293.278739&cid=C2C5FKPPB
TL;DR: bytes moved by scatter goes from lots to 0.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77403#-:-Mx_9gv389-5oWR8M2s7:b-8ckdob)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @adityamaru, and @nvanbenschoten)

___
*[-- commits, line 10 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77403#-Mx_4PeC1N3aPQOsSYJu):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> Instead, we use the next split for the first one's predicate as this can also serve to show that the span is still 'wide' enough to split.

If the span is no longer wide enough to split, won't this lead to non-empty scatters? Did you consider not using a predicate key at all in this case?
</blockquote></details>

I don't think we'll scatter at all if it turns out that we're overlapping which the hit predicate indicates, right?
___
*[pkg/kv/bulk/buffering_adder.go, line 370 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77403#-Mx_3jch7QbG7ZVhSEcM:-Mx_EBsv7_Dae1u_74iD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/036c0a47dd123cf629a99ec3983373f318f15cc1/pkg/kv/bulk/buffering_adder.go#L370)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Could you add a comment explaining this case?
</blockquote></details>

Done.
___
*[pkg/kv/bulk/buffering_adder.go, line 371 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77403#-Mx_3sMxBNdBsf1z0kzv:-Mx_ECX29Qa4lRVzXbeJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/036c0a47dd123cf629a99ec3983373f318f15cc1/pkg/kv/bulk/buffering_adder.go#L371)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: this would be easier to understand if this was written as:
```go
next := splitAt + width
if next > len(b.curBuf.entries) - 1 {
    next = len(b.curBuf.entries) - 1
}
```
</blockquote></details>

Done.
___
*[pkg/kv/bulk/buffering_adder.go, line 380 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77403#-Mx_56v25HjSbTD_Klkv:-Mx_5OIv6fvrMDo-06ng:bemgry7) ([raw file](https://github.com/cockroachdb/cockroach/blob/036c0a47dd123cf629a99ec3983373f318f15cc1/pkg/kv/bulk/buffering_adder.go#L380)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do you think we should be using a longer expiration for the first split in each partition?
</blockquote></details>

eh, after an hour we *should* have ingested enough to justify the split by size or else we don't really need it. If we're looking at true unordered data, we should have spent an hour filling all the spans and they should be big enough. If we're looking at loosely ordered data and we're filling past this split since minute 3 or whatever, then we didn't need it anyway. Either way, I think we're okay letting size decide its fate after, say, 10minutes.

**Comment by dt** [comment]
I guess if we over-split the first span in the case of loosely ordered data, then we're going to do some merging at 1h, but we'd really, really, really prefer it merge the non-first splits.

**Comment by dt** [comment]
Added an hour hour to the first split and comment explaining why. Thanks for catching that!

**Review by nvb — COMMENTED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77403)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @adityamaru, and @dt)

___
*[-- commits, line 10 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77403#-Mx_4PeC1N3aPQOsSYJu):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I don't think we'll scatter at all if it turns out that we're overlapping which the hit predicate indicates, right?
</blockquote></details>

We won't scatter, but the partition (am I using that word right?) before us will. So if we don't split directly at the start of our partition, won't the previous one perform non-empty scatters, like it always did before this change?

**Review by dt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77403)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @adityamaru, and @nvanbenschoten)

___
*[-- commits, line 10 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77403#-Mx_4PeC1N3aPQOsSYJu):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We won't scatter, but the partition (am I using that word right?) before us will. So if we don't split directly at the start of our partition, won't the previous one perform non-empty scatters, like it always did before this change?
</blockquote></details>

I'm not following: if we're looking at unsorted data, then all the processors overlap and will be adding data to many ranges for every buffer they flush; since they never flush a full "range" worth of data without hitting a range boundary, they never split and scatter anything, empty or otherwise.

**Comment by dt** [comment]
>  So if we don't split directly at the start of our partition, won't the previous one perform non-empty scatters

If our first split fails its predicate test, that means someone split between it and 1/n of our buffered keys, which indicates we don't actually have a partition (sure, works for me) of the ingestion span to ourselves; we're just all getting and adding keys all over each other, so just adding and letting ranges fill and split themselves reactively is the best we can do.

I suppose we could also remove the predicates entirely, and just let every node create a fixed number of splits like 3 or 5 or something. That'd scale by number of nodes organically (since they each do it) rather than having one node try to make all the splits if, and only if, it is first. 🤔 

In either case, I think this change -- bug fixing the first missing split in current approach -- shouldn't block on potentially changing predicates (on first or all)?

**Comment by dt** [comment]
Added another commit that fixes the fact these splits were not ensuring they were at safe split keys 🤦

**Review by nvb — APPROVED** [review]
:lgtm:

> Added another commit that fixes the fact these splits were not ensuring they were at safe split keys 🤦

What was the effect of this? Did KV detect this and reject the split attempt?

> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77403#-:-MxfaXR1-Lo8bGJ30i_U:b1ad56j)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aayushshah15, @adityamaru, and @stevendanna)

___
*[-- commits, line 10 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77403#-Mx_4PeC1N3aPQOsSYJu):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

I'm not following: if we're looking at unsorted data, then all the processors overlap and will be adding data to many ranges for every buffer they flush; since they never flush a full "range" worth of data without hitting a range boundary, they never split and scatter anything, empty or otherwise. 
</blockquote></details>

Ack, thanks for the explanation here and above.

**Comment by dt** [comment]
> What was the effect of this?

@stevendanna pointed out that we started seeing SQL errors of the form `internal error: Non-nullable column "news_item:ni_headline" with no value!"` that were mis-attributed to the new index backfill changes, but in actuality started shortly after this code was enabled when I fixed the incorrect conditional around calling `createInitialSplits`. That rang a bell based on when sstBatcher's splits made the same mistake a couple years ago.

> Did KV detect this and reject the split attempt?

I don't _think_ we actually check `EnsureSafeSplitKey` in `AdminSplit`, but maybe we should?

TFTR!

bors r+

**Comment by dt** [comment]
@stevendanna well, that bors CI just failed on the same `TestPrimaryKeyChangeWithOperations` flake we've been seeing elsewhere, i.e.  `pq: internal error: unexpected error from the vectorized engine: null_value_error: non-nullable column "test:x" with no value! Index scanned was "test_pkey" with the index key columns (y) and the values (1765)`, but that is with the addition of EnsureSafeSplitKey here, so maybe that doesn't explain that error after all and this patch isn't fixing it as we'd hoped.

That said, I still think this diff _is_ fixing the missing first split (as shown by the scattered number delta) so I don't think it is flawed, but we'll need to keep digging for the cause of that flake if this isn't fixing it too.

bors r+

**Comment by stevendanna** [comment]
@dt Yes, I think it is expected that the fix you added here doesn't address `TestPrimaryKeyChangeWithOperations`. While they have the same end error message, they are caused by different issues. I think the failure bors hit is the bug I'm fixing here: https://github.com/cockroachdb/cockroach/pull/77464
